bjoern/server.c

```diff
@@ -141,46 +196,81 @@ ev_timer_ontick(struct ev_loop* mainloop, ev_timer* watcher, const int events)
 }
 #endif
 
+#ifdef WANT_GRACEFUL_SHUTDOWN
+static void
+ev_shutdown_ontick(struct ev_loop* mainloop, ev_timer* watcher, const int events)
+{
+  ThreadInfo *thread_info = (ThreadInfo*)ev_userdata(mainloop);
+  DBG("ev_shutdown_ontick %d", thread_info->server_info->active_connections);
+
+  if (thread_info->server_info->active_connections == 0) {
+    DBG("No more active connections, shutting down");
+    GIL_LOCK(0);
+
+    ev_timer_stop(mainloop, &shutdown_watcher);
+    ev_timer_stop(mainloop, &shutdown_timeout_watcher);
+
+#ifdef WANT_SIGINT_HANDLING
+    ev_signal_stop(mainloop, &thread_info->sigint_watcher);
+#endif
+
+    GIL_UNLOCK(0);
+  } else {
+    DBG("Still have active connections %d", thread_info->server_info->active_connections);
+  }
+}
+
+static void
+ev_shutdown_timeout_ontick(struct ev_loop* mainloop, ev_timer* watcher, const int events)
+{
+  ThreadInfo *thread_info = (ThreadInfo*)ev_userdata(mainloop);
+
+  GIL_LOCK(0);
+  ev_timer_stop(mainloop, &shutdown_watcher);
+  ev_timer_stop(mainloop, &shutdown_timeout_watcher);
+  // break because we've exceeded the timeout and want to just stop where we are,
+  // regardless of where the other watchers are
+  ev_break(mainloop, EVBREAK_ALL);
+  DBG("Shutdown took too long, %d active connections", thread_info->server_info->active_connections);
+  GIL_UNLOCK(0);
+}
+#endif
+
 static void
 ev_io_on_request(struct ev_loop* mainloop, ev_io* watcher, const int events)
 {
   int client_fd;
   struct sockaddr_in sockaddr;
   socklen_t addrlen;
-
   addrlen = sizeof(struct sockaddr_in);
-  client_fd = accept(watcher->fd, (struct sockaddr*)&sockaddr, &addrlen);
-  if(client_fd < 0) {
-    DBG("Could not accept() client: errno %d", errno);
-    STATSD_INCREMENT("conn.accept.error");
-    return;
-  }
-
-  int flags = fcntl(client_fd, F_GETFL, 0);
-  if(fcntl(client_fd, F_SETFL, (flags < 0 ? 0 : flags) | O_NONBLOCK) == -1) {
-    STATSD_INCREMENT("conn.accept.error");
-    DBG("Could not set_nonblocking() client %d: errno %d", client_fd, errno);
-    return;
-  }
+  while ((client_fd = accept(watcher->fd, (struct sockaddr*)&sockaddr, &addrlen)) > 0) {
```

Comment:
Is it safe to accept multiple requests in a single callback call? If yes, suggest to rename to `ev_io_on_requests`